### PR TITLE
Improve type inference for plugin manager implementations

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Improves type inference particularly for plugin managers in dependent libs

### CODE REVIEW

**Comment by boesing** [comment]
Just a side-note here:

I was about adding this in #98 
Back then, @Ocramius told me that this would be a soft BC break due to the way how generics work. So when generic declaration is missing, this will become very messy in projects after upgrading to the version introducing generics.

Therefore, I deferred that to servicemanager v4

File: src/ServiceLocatorInterface.php [hunk 1]
```diff
@@ -15,9 +15,11 @@ interface ServiceLocatorInterface extends ContainerInterface
     /**
      * Build a service by its name, using optional options (such services are NEVER cached).
      *
-     * @param  string $name
+     * @template T of object
+     * @param  string|class-string<T> $name
      * @param  null|array<mixed>  $options
      * @return mixed
+     * @psalm-return ($name is class-string ? T : mixed)
```
Comment (Ocramius) [inline src/ServiceLocatorInterface.php:22]: In theory, yes.
In practice, this is not necessarily true.

I think adding type inference is good though, even if it **may** be a lie :thinking: 

It's an accepted convention, so I'd say :+1:, and whoever breaks the convention is at fault here instead.

File: src/ServiceManager.php [hunk 2]
```diff
@@ -259,6 +259,9 @@ public function build($name, ?array $options = null)
 
     /**
      * {@inheritDoc}
+     *
+     * @param string|class-string $name
```
Comment (Ocramius) [inline src/ServiceManager.php:263]: `string` contains `class-string` :thinking:

**Review by Ocramius — APPROVED** [review]
LGTM - should we :ship: ?

**Comment by Ocramius** [comment]
> Back then, @Ocramius told me that this would be a soft BC break due to the way how generics work. 

I think this only applies to our `latest` CI builds, where `psalm` would complain about a non-provided template parameter?

I'd say this is good improvement, and doesn't affect the API per se, just documentation. Downstream consumers that use psalm are used to upgrades "shifting" the types a bit, and will adjust.

**Comment by boesing** [comment]
> I'd say this is good improvement, and doesn't affect the API per se, just documentation. Downstream consumers that use psalm are used to upgrades "shifting" the types a bit, and will adjust.

I'm fine with this. 

Anyways: we need to adjust our own components as well, which are quite a lot.
Can we probably create a project for this and list all components which need work then?

Thinking about:
- input filters
- validators
- cache adapters
- cache plugins (or did I killed that? not 100% sure...)
- form elements
- etc.

_(btw: with `we` I mean "not me" 😄)_

**Comment by Ocramius** [comment]
> Can we probably create a project for this and list all components which need work then?

I wonder if the installation of renovate (which didn't happen yet, I know, sorry :( ) would already handle this?

**Comment by boesing** [comment]
> I wonder if the installation of renovate would already handle this?

Since I don't think that renovate will create a list of components where the update to the latest `servicemanager` version fails, I'd say no. And I doubt that all the affected projects do have `renovate` setup anyways.

**Comment by Ocramius** [comment]
Well, `psalm` would fail there, and renovate would open a PR :D

**Comment by froschdesign** [comment]
> Can we probably create a project for this and list all components which need work then?

I can create the project to track the progress and to give an overview.

**Comment by Ocramius** [comment]
@froschdesign please do: I just tried creating one, and it messed up everything, possibly because I have 2 versions of GitHub Projects on my system (the pre-release one too)

File: src/ServiceManager.php (hunk 2 above)
Comment (gsteel) [reply src/ServiceManager.php:263]: As documentation, this reads to me like - "I might get a regular string literal or a class-string" - is that not an improvement on plain old string?

Comment (Ocramius) [reply src/ServiceManager.php:263]: No strong feelings, but other tools do indeed report this as a redundant type union

File: src/ServiceLocatorInterface.php (hunk 1 above)
Comment (Ocramius) [reply src/ServiceLocatorInterface.php:22]: Yeh, let's go with it 👍

**Comment by Ocramius** [comment]
Meanwhile, :ship: here.

File: src/ServiceLocatorInterface.php (hunk 1 above)
Comment (Slamdunk) [reply src/ServiceLocatorInterface.php:22]: If we got it for `build`, that makes sense for `get` too, doesn't it?
`psr/container` would not accept it, but we can do it :wink:

**Comment by froschdesign** [comment]
@gsteel @Ocramius @boesing 
With some delay, here is a first start: https://github.com/orgs/laminas/projects/21

**Pull request merged**